core/trino-spi/src/main/java/io/trino/spi/SplitWeight.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.spi;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonValue;
+
+import java.util.Collection;
+import java.util.function.Function;
+
+import static java.lang.Math.multiplyExact;
+
+public final class SplitWeight
+{
+    private static final long UNIT_VALUE = 100;
```

Comment:
> Since using a floating point number is not an option (due to error accumulation)

@arhimondr are you sure?

Reply:
That's a great question. I'm not sure actually. That was a takeaway from a quick offline discussion we had with James. While in theory floating point numbers tend to loose precision on long running sequences or when operating on numbers that are multiple orders of magnitude different it doesn't feel like the split queue depth should result in high magnitude differences between the current queue weight and a single split weight.

```
    public static void main(String[] args)
    {
        double doubleSum = 10_000_000;
        BigDecimal bigDecimalSum = BigDecimal.valueOf(10_000_000);
        for (int i = 0; i < 100_000_000; i++) {
            doubleSum += 0.000001;
            bigDecimalSum = bigDecimalSum.add(BigDecimal.valueOf(1, 6));
        }
        System.out.printf("Double: %s, Decimal: %s\n", doubleSum, bigDecimalSum);
        for (int i = 0; i < 100_000_000; i++) {
            doubleSum -= 0.000001;
            bigDecimalSum = bigDecimalSum.subtract(BigDecimal.valueOf(1, 6));
        }
        System.out.printf("Double: %s, Decimal: %s\n", doubleSum, bigDecimalSum);
    }
    
    > Double: 1.0000100024044514E7, Decimal: 10000100.000000
    > Double: 1.0E7, Decimal: 10000000.000000
```

I wonder if the precision lose is significant for this specific use-case? @pettyjamesm could you please elaborate a little more on problems you observed with `double`?

Reply:
I didn't specifically experiment with, and then run into issues with `double`- but rather discarded it because of the potential for these problems. A `SplitWeight` implementation using double that didn't enforce anything beyond values being > 0 would run the risk of accumulating rounding error either on the task side or coordinator side total weight accounting (either for node total or task queued split weights) wouldn't balance to 0. If that were to happen, then the fractional error could either let the weight value dip below 0.0 or could stay above 0 even after all splits finished which would prevent that node from being fully utilized in the future because of a phantom weight.

It was easier to constrain this implementation to integer semantics to make the accounting sane, but with the unfortunate consequence of having to pick some arbitrary value as the "standard". I talked to @arhimondr about this offline today and I think there's a balance point that we can reach to make weights "look" more like values relative to `1.0` for the ergonomics of the API but still use integer arithmetic for the internal book keeping logic.

Reply:
Currently the weight is tracked per node and per task. 

Realistically it doesn't feel like like an error can accumulate significantly in a scope of a task. The maximum weight assigned shouldn't exceed 100 - 1000 with the minimum weight being in the 0.01 - 0.0001 range. The space of variation for weight in this scenario shouldn't exceed ~8 decimal digits. Double uses 54 bits to express values (that is ~16 decimal points) leaving another 8 decimal points for rounding errors. 

While in theory it is possible to come up with a sequence of rounding errors that would push error accumulation over `10^-8` this sequence will be long (probably much longer than the total number of splits assigned to a single task) and rather unlikely. The weight value is also going to be discarded as soon as the task is finished thus it should be okay if it doesn't converge to 0 at the end.

From what I understand the node level weight tracking is a concern here. It could potentially have a much higher variation space leaving fever "digits" for rounding errors and much much longer sequences of arithmetic operations (as long as a total number of splits assigned to a node during it's lifetime). It feels like it is safer to encode weight as an integer, @findepi what do you think?

Reply:
I was answering the original problem that picking integers fixes the minimal granularity we operate at. I didn't follow how these weights are used, and whether the error can accumulate over time, and what are the consequences of this being the case. Would the system no longer be 'perfectly fair', or would it start doing nonsense?